toolchain/parse/extract.cpp

```diff
@@ -36,43 +36,56 @@ namespace Carbon::Parse {
 template <typename T>
 struct Extractable;
 
+struct Tree::ExtractState {
```

Comment:
This structure seems to be uncommented -- I think it'd be good to explain what it is and how it's used.

But also, would this make sense as a class with the various functions that are passing around `ExtractState` instead as member functions? You already have several member functions here, and it doesn't feel as struct-y as a result.

Reply:
Documented and made some things members. Given the way the `Extractable` specializations work out, keeping the members public seems a little simpler than moving all the remaining code into member functions and turning this into a class, but I can do that if you prefer.

Reply:
I'd really prefer a class here. For the Extractable specializations, would a `template <typename T> friend Struct Extractable<T>;` avoid any accessor overhead? (though alternately, it looks like trace and node_id might be the only things on state that are access?)

Reply:
Made into a class. Could probably make `Extractable` be a member class, but so far that doesn't seem necessary; using the public API seems to be working fine.